OpenSim/Simulation/MarkersReference.cpp

```diff
@@ -24,133 +24,168 @@
 #include "MarkersReference.h"
 #include <OpenSim/Common/Units.h>
 #include <SimTKcommon/internal/State.h>
+#include <cmath>
 
 using namespace std;
 using namespace SimTK;
 
 namespace OpenSim {
 
-MarkersReference::MarkersReference() : Reference_<SimTK::Vec3>()
-{
+MarkersReference::MarkersReference() :
+    Reference_<SimTK::Vec3>() {
     constructProperties();
     setAuthors("Ajay Seth");
 }
 
 MarkersReference::MarkersReference(const std::string& markerFile,
-    Units modelUnits) : MarkersReference()
-{
+                                   Units modelUnits) :
+    MarkersReference() {
     loadMarkersFile(markerFile, modelUnits);
 }
 
-
-MarkersReference::MarkersReference(MarkerData *markerData,
-    const Set<MarkerWeight>* markerWeightSet) : MarkersReference()
-{
-    _markerData.reset(markerData);
-    if (markerWeightSet!=nullptr) upd_marker_weights()= *markerWeightSet;
-    populateFromMarkerData(*markerData);
+MarkersReference::
+MarkersReference(const TimeSeriesTable_<SimTK::Vec3>& markerTable,
+                 const Set<MarkerWeight>* markerWeightSet) :
+    MarkersReference() {
+    _markerTable = markerTable;
+    if(markerWeightSet != nullptr)
+        upd_marker_weights() = *markerWeightSet;
+    populateFromMarkerData(markerTable);
 }
 
-/** load the marker data for this MarkersReference from markerFile  */
-void MarkersReference::loadMarkersFile(const std::string markerFile, Units modelUnits)
-{
-    upd_marker_file() = markerFile;
-    _markerData.reset(new MarkerData(get_marker_file()));
+void MarkersReference::loadMarkersFile(const std::string markerFile,
+                                       Units modelUnits) {
+    auto fileExt = FileAdapter::findExtension(markerFile);
+    OPENSIM_THROW_IF(!(fileExt == "sto" || fileExt == "trc"),
+                     UnsupportedFileType,
+                     markerFile,
+                     "Supported file types are -- STO, TRC.");
+
+    if(fileExt == "trc") {
+        _markerTable = TimeSeriesTableVec3{markerFile};
+    } else {
+        try {
+            _markerTable = (TimeSeriesTable{markerFile}).pack<SimTK::Vec3>();
+        } catch(const IncorrectTableType&) {
+            _markerTable = TimeSeriesTable_<SimTK::Vec3>{markerFile};
+        }
+    }
+    if(_markerTable.hasTableMetaDataKey("Units"))
+        _units = Units{_markerTable.getTableMetaData<std::string>("Units")};
```

Comment:
> Units are handled correctly if we load from file. Can we abstract this conversion and use it at the time the reference value is requested. Then it would not matter how the Table was supplied (directly or by file).

If Units is known during construction of MarkersReference, the quantities can be converted during construction and the TimeSeriesTable can hold this information as metadata. Moving this conversion process to `getValues()` will require the function to know the *current* and *target*.

> Minor, Units is a local helper and does not need to be data member.

I have removed the data-member.